master/internal/api_user.go

```diff
@@ -683,3 +696,199 @@ func (a *apiServer) PostUserActivity(
+}
+
+// POST User Long lived token takes user id and optional lifespan and overwrites an
+// access token for the given user.
+func (a *apiServer) PostUserLongLivedToken(
+	ctx context.Context, req *apiv1.PostUserLongLivedTokenRequest,
+) (*apiv1.PostUserLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = user.AuthZProvider.Get().CanCreateUsersToken(ctx, *curUser, targetUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := user.DefaultTokenLifespan
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, err := user.DeleteAndCreateLongLivedToken(
+		ctx, targetFullUser.ID, user.WithTokenExpiry(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostUserLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+// GET Long lived token returns token information for the logged in user.
+func (a *apiServer) GetLongLivedToken(
+	ctx context.Context, req *apiv1.GetLongLivedTokenRequest,
+) (*apiv1.GetLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err = user.AuthZProvider.Get().CanGetUsersOwnToken(ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenInfo, err := user.GetLongLivedTokenInfo(ctx, curUser.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiv1.GetLongLivedTokenResponse{LongLivedTokenInfo: tokenInfo.Proto()}, nil
+}
+
+// GET All Long lived token returns all long-lived token information from user_session db.
+func (a *apiServer) GetAllLongLivedToken(
+	ctx context.Context, req *apiv1.GetAllLongLivedTokenRequest,
+) (*apiv1.GetAllLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// targetFullUser, err := getFullModelUser(ctx, curUser.ID)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// targetUser := targetFullUser.ToUser()
+
+	if err = user.AuthZProvider.Get().CanGetAllToken(ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenInfo, err := user.GetAllLongLivedTokenInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var res []*userv1.UserSessionInfo
+	for _, s := range tokenInfo {
+		res = append(res, s.Proto())
+	}
+	return &apiv1.GetAllLongLivedTokenResponse{LongLivedTokenInfo: res}, nil
+}
+
+// GET User Long lived token takes user id and returns token information for the given user.
+func (a *apiServer) GetUserLongLivedToken(
```

Comment:
question: Do we still have a use case for this endpoint? I don't think we currently do since the plan would be to use `GetAllLongLivedTokens` for the admin list tab.

Reply:
My 2cents: I believe this API would be beneficial because there could be a need to retrieve token for a specific user. Searching for a particular token by navigating through pages of results from the `GetAllLongLivedTokens` API could be cumbersome and time-consuming.

But I am completely open to suggestions

Reply:
You make a good point, without filtering & sorting, navigating through the token management table will get super tedious! It might be worth adding in that functionality to that endpoint as well to more closely match something like the `/api/v1/users` endpoint
```
/api/v1/users?sortBy=SORT_BY_MODIFIED_TIME&orderBy=ORDER_BY_DESC&offset=0&limit=10&name=a
```

So for `/api/v1/user/tokens`, we could potentially do something like:
```
/api/v1/user/tokens?sortBy=SORT_BY_EXPIRY&orderBy=ORDER_BY_DESC&offset=0&limit=10&description=a
```
We could also potentially include a way to filter based on `user_id`/`user_name`'s

Reply:
I'm trying to think of the case where we'd want to look up the token for a specific user from an admin's perspective outside if using the `GetAllLongLivedTokens` workflow. By no means am I saying there's not one, I just don't see it right now haha

Reply:
But you point absolutely shows a clear need to have proper filtering & sorting on this table especially if this features is widely adopted

Reply:
Keeping this conversation active. 
I am keeping `GetUserLongLivedToken` till I get to developing the CLI for GET APIs. :D 
Added `SORT` and `FILTER` to GET All API, with a filter on `username`

Reply:
Finally after having completed CLIs, I can attest that we would be needing all the 3 GetLongLivedToken APIs.  
```
GetLongLivedToken // GET active Long lived token returns token information for the logged in user.
```
We need this when running `det user token create ` CLI command (without username)
```
GetAllLongLivedTokens // GET All (active / revoked) Long lived token returns all long-lived token information from user_session db.
```
We need this when running `det user token ls --all  ` CLI command 
```
GetUserLongLivedToken // GET User Long lived token takes user id and returns token information for the given user.
```
We need this when running `det user token create --username ` CLI command

Reply:
we were able to unify them, and changes are pushed, so resolving the conversation